packages/server/modules/core/services/ratelimiter.ts

```diff
@@ -0,0 +1,184 @@
+    limitCount: getIntFromEnv('RATELIMIT_POST_OBJECTS_STREAMID', '200'),
+    duration: 1 * TIME.minute
+  },
+  'GET /objects/:streamId/:objectId': {
+    limitCount: getIntFromEnv('RATELIMIT_GET_OBJECTS_STREAMID_OBJECTID', '200'),
+    duration: 1 * TIME.minute
+  },
+  'GET /objects/:streamId/:objectId/single': {
+    limitCount: getIntFromEnv('RATELIMIT_GET_OBJECTS_STREAMID_OBJECTID_SINGLE', '200'),
+    duration: 1 * TIME.minute
+  }
+}
+
+const redisClient = new Redis(getRedisUrl(), {
+  enableReadyCheck: false,
+  maxRetriesPerRequest: null
+})
+
+export const sendRateLimitResponse = (
+  res: express.Response,
+  action: string,
+  rateLimiterRes: RateLimiterRes | undefined
+): express.Response => {
+  if (rateLimiterRes instanceof Error) {
+    Sentry.captureException(rateLimiterRes)
+    res.setHeader('X-Speckle-Meditation', 'https://http.cat/500')
+    return res.status(500).send({
+      err: 'Error when attempting to determine rate limit. Please try again later.'
+    })
+  }
+  if (rateLimiterRes) {
+    res.setHeader('Retry-After', rateLimiterRes.msBeforeNext / 1000)
+    res.setHeader('X-RateLimit-Remaining', rateLimiterRes.remainingPoints)
+    res.setHeader(
+      'X-RateLimit-Reset',
+      new Date(Date.now() + rateLimiterRes.msBeforeNext).toISOString()
+    )
+  }
+  if (action) {
+    const opts = LIMITS[action]
+    res.setHeader('X-RateLimit-Limit', opts.limitCount)
+  }
+
+  res.setHeader('X-Speckle-Meditation', 'https://http.cat/429')
+  return res.status(429).send({
+    err: 'You are sending too many requests. You have been rate limited. Please try again later.'
+  }) // TODO we should return a branded page (either here, or via nginx)
+}
+
+export const rateLimiterMiddleware = async (
+  req: express.Request,
+  res: express.Response,
+  next: express.NextFunction
+) => {
+  if (isTestEnv()) next()
+
+  const burstAction = 'ALL_REQUESTS_BURST'
+  const action = 'ALL_REQUESTS'
+
+  const source: string = (req as RequestWithContext)?.context?.userId
+    ? ((req as RequestWithContext)?.context?.userId as string)
+    : getIpFromRequest(req)
+
+  isWithinRateLimits({ action, source })
+    .catch((rateLimiterResponse) => {
+      sendRateLimitResponse(res, action, rateLimiterResponse)
+    })
+    .then(() => isWithinRateLimits({ action: burstAction, source }))
+    .catch((rateLimiterResponse) => {
+      sendRateLimitResponse(res, burstAction, rateLimiterResponse)
+    })
+    .then(() => next())
+}
+
+// Promise will reject if the source is not within limits for the action, resolve otherwise
+export async function isWithinRateLimits({
+  action,
+  source
+}: isWithinRateLimitsConfig): Promise<RateLimiterRes> {
+  const rlOpts = LIMITS[action]
+  if (!rlOpts) return Promise.reject(null) // the rate limits for the action have not been defined, so prevent use of the action
+
+  // provides a backup store for the rate limiter
+  // incase Redis is unreachable
+  const rateLimiterMemory = new RateLimiterMemory({
```

Comment:
you are creating a new instance of the memory rate limiter for every call of this function. 

That doesn't really do anything, that instance needs to be created once at startup.

Possibly its also inefficient to create a new instance of the RateLimiterRedis for earch request